### PR TITLE
HD SPESSMEN: 64x64 take two

### PR DESCRIPTION
**Pull request by ShiftyRail**
# About this PR

This PR tries to bring back a forgotten project form 2016 to upscale the game to 64x64.
Most of the work was done by PJB3005, Clusterfack, and N3X15 with ByondTools.

I only had to apply their work and yell insults at Python scripts until they behaved correctly.

## But why?

- Allows our talented spriters to sprite with up to 4 times the detail
- Looks pretty rad and crispy

## It didn't work then, why would it work now?

Byond evolved quite a bit during the last 4 years. Most of the issues 64x64 caused are now fixed, byond-side.
- **Compile times are no longer horrendously long.** My copy of /vg/ compiles in less than a minute with or without 64x64.
- **Better maptools than DM exist.** 64x64 completely *kills* DreamMaker as a mapping tool, there's simply too much pixels in our already bloated maps. Fortunately, better map tools exist now, so now is the time to ditch DM and switch to them.
- **Sprite corruption is less of a thing.** Back in 2016 or 2017, it was relatively frequent to have a corrupted `.rsc` which resulted in random sprites being swapped together, ie, a fragment of a spacepod laid over a space tile, and other things like that. 64x64 worsened this problem. This was mostly a thing in BYOND 510, which we used for much longer than we had to because of our screen-freeze bug. BYOND 511 was much less prone to this bug, and I've frankly never seen it in BYOND 512.

## What still needs to be done ?

- [x] Upscale lobby screens (they are .gif or .png, I'll find a way)
- [x] Find & fix all maptext outputs
- [ ] Find a way to make the statpanel icons less XBOGHUEG (I hope there's a way to do this with skin fuckery, but I would need to use more brain power than I have).
- [x] Explicit calls to `pixel_whatever` hunt
- [ ] A fuckton of playtesting
- [ ] Gonflicts
- [x] HuD
- [x] A clean revert in case things go sour

## What changes concretely for me as a player?

In a first time, nothing; The game will look the exact same as now. 
Hopefully, our very talented spriters will use it for better sprites and give us a nice, HD game.

## Will there be a clash between upscaled 32x32 sprites and new, fresh 64x64 ones?

It is a risk. I don't think it is as much of a danger as people made it out to be in 2016.
Even if no spritework is done at all, there are some benefits to upscaling which are purely on the code-side, notably how transform matrices are handled or with maptext.

# Gallery

image

display_case
64 vs 32

capgun

83251237-9e8ff180-a1a9-11ea-9a51-98dabe6b9475

### CODE REVIEW

**Comment by SonixApache** [comment]
bless you and your attempts

**Comment by aceedex** [comment]
How many times must we tech you this lesson old man

**Comment by Nickvr628** [comment]
Can we get some side by side shots of what it looks like to players?

**Comment by SonixApache** [comment]
everything just gets its resolution doubled, so it looks identical

**Comment by drawsstuff** [comment]
> Can we get some side by side shots of what it looks like to players?

There is no difference, not at first, anyways. All this does is pave the road for the future.

**Comment by ghost** [comment]
> 
> 
> Can we get some side by side shots of what it looks like to players?

Everything looks identical. This doesn't actually change any icons, it just nearest neighbor scales them to 64x64. The idea is to make it easier to introduce new sprites that are higher resolution.

**Comment by streaki** [comment]
This PR seems to be a case of best intentions paving a road to hell, pardon the pun.
I believe it would only work if it came with a complete rework of all sprites in the game (which is not feasible) because otherwise it's going to lead to inconsistent visuals. I can guarantee most people are going to want to add high-res sprites, but old sprites will stay in majority practically forever, making the new ones stick out like a sore thumb. The best comparison I can come up with would be comparing the effect to a Unity asset flip.

**Comment by ghost** [comment]
I'd save the fearmongering for when sprite terrorists start making eye searing mixel PRs to blind us, making assumptions about what's going to happen is purely speculative at this point. Remember that this PR doesn't visually change anything and individual sprite PRs taking advantage of it have to pass on their own.

**Comment by aceedex** [comment]
You use this looks the same player side ? I remember some sprites were upscaled to 64x64 making them huge. The taser in particular I remember getting huge

**Comment by ShiftyRail** [comment]
@aceedex I did not see major clash in size at all.

**Comment by ghost** [comment]
I'm giving it both an upboat and a downlow because i like the idea of having smaller sprites be more detailed but at the same time i'm scared of the consequences.

**Comment by Kurfursten** [comment]
@streaki For all the fear of mixels, let's not forget that we already use 64x64 for things currently. Deity had a big habit of using it, so
>cult effects
>new blob (and related meteors)
>mannequins
>throne
>station map (the big oval tactical one)

I'm surprised bees weren't 64x64

**Comment by ShiftyRail** [comment]
image

All security items, upscaled.

**Comment by Nickvr628** [comment]
IDK it looks wrong somehow.

**Comment by ShiftyRail** [comment]
The only "noticeable" difference is that the turf lines are more crisp. That's probably what you're seeing without realising.

**Comment by GreatBigFailure** [comment]
Might as well try. 
If it works we get a drip feed of high quality sprites as spriters get to work with their 4x larger canvas. If it doesn't work we get 1-3 days of funny goofy glitches, revert it to what it is now, and make Shifty sad.
It's win win.

**Comment by ShiftyRail** [comment]
image

One of the very easy to do, noticeable, and non-clashing improvement you can do with 64x64

**Comment by Exxion** [comment]
@Kurfursten The current 64*64 sprites are that many pixels because they're double-size. All sprites in the game have the same pixel size.

**Comment by JTGSZ** [comment]
I hope this goes well, even tho I'm going to have to upscale everything Ive done lol.

**Comment by kilozombie** [comment]
>Better maptools than DM exist. 64x64 completely kills DreamMaker as a mapping tool, there's simply too much pixels in our already bloated maps. Fortunately, better map tools exist now, so now is the time to ditch DM and switch to them.

The gargantuan benefit to using DM as a mapping tool is the fact that the coding and spriting tools are right there alongside, and vice versa. Being able to see what your sprite looks like ingame, being able to examine what's in individual maps-- these are important features.

Also, while I'm in support of a move to a better mapping client (Lamprey still gives me nightmares, to say nothing of Deadharold's tortured soul), you need to rewrite the existing mapping guide to be compatible with the preferred client.

Spriting will also be a more convoluted process. I've imported in sprites from paintdotnet but it's an inefficient ordeal and difficult to get transparency working, so I know that some people sprite directly in DM. This makes designing 32x32 sprites much more annoying because you need pixels twice as big, with the alternative being spriting in 64x64, which is massively more difficult for an artist. Yes, it's 4x the freedom, but it's 4x less leeway-- the advantage of having smaller sprites is you can suspend disbelief for the shape of something a lot more, and spriting more detail on a larger canvas requires tons more work and skill.

I suspect that 64x64 sprites will look like shit ingame on smaller monitors. Right now, 1x scaling requires 480 pixels of height, but it will suddenly require 960, and when factoring in the margins of Dreamseeker and (on Win7 at least) the Windows taskbar, it seems likely to me that monitors with 1080 vertical resolution will be forced to crunch the viewport to an annoying degree to be able to see the full height of spess, with 64x64 sprites being downscaled.

There's no 'right time' to do this, only the point at which people are willing to put in the legwork on making it a smooth transition. It's overall a great move for spess, but this is going to require more work and, like you said, a fuckload of playtesting. I honestly hope this goes well but there will surely be compromises.

**Comment by ShiftyRail** [comment]
You will be able to work and notably sprite in DM. Mapping is just going to be a nightmare (it is on my craptop) because you will be forced to zoom out one more level, with all the associated lag. DM got a lot better at spritemaking than in 2016, notably, it will keep the alpha of your upscaled images. (It didn't in the past).

As far as resolution goes, my Win7 1080 monitor has no trouble displaying the game at all in fullscreen. The only technical issue (emojicracy notwithstanding) which needs to be adressed merging is the alt-click menu. It is too big.

I think there's a way to make it better, but I still need to figure it out.

**Comment by ShiftyRail** [comment]
@aceedex 

image

image

I think this is what you were talking about when thinking about "XBOG HUEG" tasers. 64x64 make tasers look nearly identical when put in a recharger, instead of losing a few pixels here and there in 32x32.

**Comment by ghost** [comment]
unironically should do a test merge, if it's even possible for a project as big as this is without breaking things.

**Comment by ShiftyRail** [comment]
@small-boss I fully understand how risky the thing is. Fist, it isn't ready yet, and when it is ready I'll make sure it is revertable without breaking anything, and I'll give a handy guide to future proof it and revert it cleanly.

Upscaling with zero upgrade to nearest neighbor allows everyone to downscale it with zero loss. I just need to write a trivial Python script for it.

I think it worth noting that you ABSOLUTELY DO NOT NEED to even rework sprites for it to be useful. 64x64 upscaling is a gain IN ITSELF, without ANY work from spriters, simply because you can get better transform matrices.

image

**Comment by SonixApache** [comment]
Hey so what do the viewport zoom modes look like with this?
ie. the `X1, X2, X4, X8` options instead of the stretch-to-view option.

**Comment by kilozombie** [comment]
> Hey so what do the viewport zoom modes look like with this?
> ie. the `X1, X2, X4, X8` options instead of the stretch-to-view option.

I brought the 1x scaling thing up because my monitor can't handle 2x (the top and bottom are cropped) and 1x is way too small. Makes me think that at 64x, 1x will be the same story as 2x, and stretch-to-fit will behave exactly like stretch-to-fit does right now.

**Comment by Exxion** [comment]
Yes, that's correct. Those scale factors are just what they sound like: The number of actual pixels on your screen per pixel in the game.
For the record, this means that anyone playing on a screen below about 1080p would be playing with downscaled sprites after this. Not that anyone still has a screen below 1080p. (The actual resolution required to display the map at full size in 64*64 is 960 square, but the title bar and taskbar almost exactly take up the rest vertically.)

**Comment by Partheo** [comment]
I'm a spriter. I have made SS13 sprites in 64x before. Let me warn you.

The code benefits are indeed interesting, can't speak to those. I can see how mapping would be improved. But let's not kid ourselves with the idea that "a drip feed of high quality sprites" is anything but an abortion of creative consistency.

Here is a stunbaton in 32x, and a stunbaton in 64x:
image
Looks cool, aye? It does, but I'd like you to use this example to reflect for a moment on how much more work is going to have to go into every single 64x sprite going forward. 32x *works*, and it is fairly accessible to those who wish to dabble in spriting. These wonderful contributors keep the game alive.  Now, if the canvas is four times bigger, the potential for detail is much, much higher, and the bar for entry is raised considerably. Making sprites this size is *hard*. You will be making spriting as a whole less accessible. This means, undeniably, that **you will have less spriters contributing moving forward**. Alternatively, you can continue to still accept 32x sprites, which brings me to my next point.

Mixels are a serious crime, folks. Refer to creative consistency mentioned earlier. Here is what a 64x sprite looks like next to upscaled 32x. I used my own image because OP can't take a proper screenshot in a discussion about the importance of resolution.
image 
image
Unless, of course, that screenshot is actually how this PR looks like in-game, in which case, feast your eyes:
image
Sore thumb is an understatement. 

Next, do not conflate "high resolution" with "high quality." An artist of the some talent will not be able to make sprites that look any better just because they have a bigger canvas. More pixels requires more work, more talent, and given the general level of competence in the spriting community, you will end up with a lot of sprites that look bad, except now they look bad in HD.
image
There is *ample* work that can be done to improve the current sprites without doubling the resolution of what we have now. We're not being held back.

My last point about consistency will be about standards- currently, if one desires to sprite an object, they can look to other sprites to get a general idea of what to do with theirs. How long do I make my gun? What sort of detail do I use for this panel? How thick do I make this rod? Etc. With 64x, there is a lot more room for variance, and there are no other sprites you can look to for styling. The potential for varying styles is significantly increased, so your first(read: all) waves of 64x sprites are going to be a bit all over the place.

Now, if you could magically redesign every single sprite in the game to look good in 64x, then actually make those sprites, and then add those to the game, I think this could be a good idea. There is a lot more potential in 64x. But that is a *mind-boggling* amount of work. Simply changing the current SS13 human, which looks absolutely horrid and has dozens of angry spriters who want it gone, is a near impossible project due to the insane amount of work required- clothing, inhands, limbs, etc. And that's all within 32x. Changing even "the majority" of 32x sprites into 64x is straight up unfeasible, and even with dedicated and focused spriters who are getting paid it would take you years to get anything that is starting to look consistent. Unless there is a genuine plan to do this, you are opening the door to the mixel apocalypse which will plague you for years to come, probably forever.

I leave you with a treat from the pixelart discord, mixels.png
mixels

**Comment by Exxion** [comment]
Yes, this is
* Bad for what little consistency we have
* Bad for support of older computers
* Bad for support of slower internet connections
* Bad for accessibility of contributing new things

Mostly for the benefit of making text and a few tiny icons look better. It's frankly baffling it has so much support.

**Comment by RaccoonBro** [comment]
How would it be bad for slow internet connections? The /icon folder only jumps up about 25MB over the base folder.

**Comment by Dacendeth** [comment]
@Partheo your stun baton looks like a lamp post you might want to get that checked out.

**Comment by Exxion** [comment]
Yes, that's 25 more megabytes\* for every player to download when they connect every time any resource file in the codebase is updated

\*The actual change might be a bit less than this due to the whole rsc being zipped, but probably not much less because PNG is already compressed and a more general compression algorithm on top of that isn't going to do much better

**Comment by RaccoonBro** [comment]
You use more than 25mb jerking off to hentai twice a day. Going from 60MB~ to 90MB~ won't be any noticeable difference to players.

in the lord's year 2020 if you can't spare a 25mb download once every few days that's a you problem tbh

**Comment by Exxion** [comment]
You're forgetting that it's downloaded through BYOND, which doesn't tell you it's downloading it unless you know where to look, and instead just nukes your ping to the server for the duration of the download
This is why you get people in the thread posting screenshots of five-digit pings

**Comment by Partheo** [comment]
> 
> 
> @Partheo your stun baton looks like a lamp post you might want to get that checked out.

See what I mean? Even a sprite designed faithfully to the reference looks way off, by simple virtue of being different. Take a look:
image
Stunbaton

It is much closer to the reference than the default stunbaton sprite, but now in higher resolution there is even more room for lack of clarity. Bigger is *not* always better, in this case.
why

**Comment by ShiftyRail** [comment]
@Exxion 

>Bad for what little consistency we have
>Bad for accessibility of contributing new things

As for all the fearmongering and posting an hilariously oversized baton which is clashing with the rest of the sprite because it is MEANT to clash with the rest of the sprites, none of this **has** to happen. You could very easily just decide most "in world" sprites have to follow a "identical to 32x32 rule", with very small things like holobadges or medals or anything being exempted.

>bad for older computers

who can't at least use 960px of resolution in 2020? Using strech-to-fit for my 1900x1080 laptop from goddamn 2013, I have zero noticeable difference than using "pure" 64x64. It still looks okay even when I'm in windowed mode provided I don't use byond "blur" mode.

>bad for slow connections

It's correct and it's something I can't test locally. If it's unbearable, then we'll simply revert this PR and call it a day, this means byond is unable to handle it. Again.

>baffling how many people support it

Almost as surprising as the amount of people who are invoking some phantom ghost of an hypothetical contributor ruining the game in a single stroke. Nobody PR'd this awful baton yet, and if someone did, people WOULD downvote it.
I think the guys upvoting are just willing to give it a try and see how it would actually play out and how it would actually feel in game. It's risky but I think it *is* worth trying, especially with the improvements to maptext and transform matrices.

@Partheo as said in the first part of my response, there's plenty of improvement that can be done without changing the general look of the game. I really like the improvements to guns in chargers or in display cases, or for small items.

scrnshot2

scrnshot3

This is not clashing, pretty easy to do, impossible in 32x32,. When I think of the potential for 64x64, I think of this, not of suddenly putting an oversized baton in the game.

**Comment by Exxion** [comment]
I didn't say it would ruin the game or anything. I said it has several costs for a very small benefit.
It's not an awful idea. Just not a good one.

**Comment by ShiftyRail** [comment]
That's fair, but I really think it has non-trivial benefits. 
Thanks for saying it in a way that isn't showing hilariously off sprites in bad faith.

I'll try to fix conflicts, make this in a mergeable state, and leave the branch open in my repo if anyone wants to try it in the future, because I don't want this to be merge begrundingly with people waiting for it to fail & say I told you so.

**Comment by Partheo** [comment]
First off, I think this looks terrible. It just does.
image

> This is not clashing

It absolutely is.

Resizing sprites should not be done, 32x or 64x otherwise. Also, riddle me this- what happens once the taser has a 64x sprite? Guns are some of the sprites that would benefit the most from the increased resolution, as you could get more realistic proportions. What happens when you put it in the charger? Same exact issue, unless you want to upscale the game to 128x to combat that issue when it comes up?



Taking a gun sprite and pasting it on top of the recharger is stupid, and that's without considering the mixels. The correct answer to this artistic problem is to make states for the chargers when they are recharging a taser/lasergun/etc. Remove the resizing problem entirely and sprite it like a man.

>  impossible in 32x32

Here.

recharge



> saying it in a way that isn't showing hilariously off sprites in bad faith

I apologize if the humor of my write-up came off as offensive, but I am trying to give actual advice here. I'm one of the many contributors who actually make sprites for this game, and my genuine opinion is that this is a bad idea.

The only way this wouldn't end up in stylistic degradation is if you somehow replaced all the sprites with 64x variants (not going to happen) or if you remove any possible case of mixelation. The latter means not accepting any sprites that have a resolution above 32x, and getting rid of any and all resizing code, like I showed above. Resizing any sprite in relation to its neighbors, for any reason, is a creative crime. If that is respected, then, sure- add your extra MB to the repo if it means better code.

Lastly, it amuses me greatly that the stunbaton is not only being called out for being awful, but being pointed to as an example of bad faith. It's literally a faithful representation of the original reference that the baton is inspired by. It is incredibly ironic that you are calling it "hilariously off". Yes. *That's the point.*

You complained it was oversized, well, here it is shortened to the same size as the other baton.
image
Looks a bit squat suddenly doesn't it? 64x completely changes your perspective with regards to proportions on sprites. Unless you want everything looking like a chode, lots of new objects are going to look oversized- not because they're actually bigger, but because that's how proportions work- you could make a much taller looking human, for example, in the same space now.

**Comment by ShiftyRail** [comment]
It does look off when zoomed in. What matters is how it is going to look *in-game*, and that's more or less the intent of my screenshots, one of them in native 64x64 and another in stretch-to-fit using slightly less pixels to display the 960 needed. 
I think it looks fine in-game.

I don't really understand the second picture. "Impossible to do in 32x32" was relating to the holobadge thing, which, I believe, look better now that more space is given to the blue thing rather than the red outline.

>Resizing sprites should not be done, 32x or 64x otherwise

This already happens in the game world. The cloning tubes are resizing humans to fit in them, rechargers, obviously resize the gun they are recharging, and display cases just shrink everything to 19x19 (i have no idea why such an odd number was chosen) in order to fit in them.

image

This shows a direct comparison of the effect of resizing in a recharger in 32x stretch-to-fit, 64x, and 32x being "forced" to render as a 2x. You might argue that both look bad, *but* the 64x one at least doesn't blur the caplaser to make it fit. 

>I apologize if the humor of my write-up came off as offensive,

Not really as offensive, more as just weird, disconnected, and wildly hypothetical when no such sprite has been merged yet. You're clearly right when you demonstrate that a pure 64x64 baton alone clearly doesn't belong in our game world, but I believe everyone knew that already.
However, a net benefit, from my point of view, of this PR is that it allows the holobadge (I use this example because I believe it stands out pretty well) to be displayed better on the officer wearing it, be it in stretch-to-fit with BYOND's blur or in "pure" 64x64.

holobadge self

holobadge self nonstretch

As an aside, there's much less blur when using BYOND's stretch-to-fit thing in the 64x64 world.

**Comment by Kurfursten** [comment]
Spriting everything to have specific sprites for every possible object with a cell is inelegant and poorly maintainable from a coding perspective. You sound like you know what you're talking about from an art point of view but that's absolutely ugly as a coder.

**Comment by ShiftyRail** [comment]
Oh, I finally get what you meant with that gif of the upside-down taser. It's neat, but it doesn't give out any info on the taser itself, to the untrained eye and with a normal in-game zoom there's zero telling that it is a taser you are recharging.

**Comment by Partheo** [comment]
> I think it looks fine in-game.
> You might argue that both look bad
> This already happens. The cloning tubes are resizing humans to fit in them, rechargers,

Fair, the effect is quite subtle I admit. And yes, I am arguing that both look bad. I think resizing a sprite for any reason is bad. The cloning tubes are also something that looks off.

> I don't really understand the second picture

It's a gif in case it didn't play properly- I was showing an example of how you can remove the resizing issue entirely by just using proper sprites in the 32x space. Clever spriting solves many clarity issues that others might think can only be solved with code.

>However, a net benefit, from my point of view, of this PR is that it allows the holobadge (I use this example because I believe it stands out pretty well) to be displayed better on the officer wearing it, be it in stretch-to-fit with BYOND's blur or in "pure" 64x64.

image
I'm not quite sure what I'm looking at here. The top image is really blurry so I can't tell what's going on. Overall the second image looks good to me. Can you elaborate on what benefit is gained in regards to the holobadge?

>Spriting everything to have specific sprites for every possible object with a cell is inelegant 

Sorry, but I have to disagree. Don't tell me it's inelegant when the alternative is literally resizing the sprite and pasting it on top of another.

>and poorly maintainable from a coding perspective. 

The "specific sprites for every possible ____" argument falls flat when you consider that SS13 is built on these. Inhands are hand drawn. Directional sprites are hand drawn. All the jumpsuit variants, lockers, clothes, etc that come in all sorts of colors are hand drawn. Spriting each object separately is a lot of work but it is the only way that gives full creative freedom. This is a spriter's burden, not a coder's, it has been maintainable in the past and still is.

>  that's absolutely ugly as a coder.

Sure, if you prefer resized sprites to hand drawn custom states. Up to you.

> but it doesn't give out any info on the taser itself, to the untrained eye

This would only be a problem if "untrained eyes" existed. It takes a single moment to make the connection between a sprite and what it represents, assuming the sprite is alright. There's zero indication that the taser is what it is, a taser. It has a vague gun shape and some lights. There's also zero indication that the recharger is a recharger- it just looks like some sort of alien pedestal. But that isn't a problem, is it? 

> there's zero telling that it is a taser you are recharging.

It's the same shape as the taser, so...

**Comment by ShiftyRail** [comment]
I play sec a fair bit and I can assure that, at normal zoom, I would probably be unable to see the thing is a taser since the only thing which sticks out of the recharger is the stick of the gun.

At a glance, it's hard to tell. Your sprite also hides, I think, how charged the gun is. It's something that is already hard to tell on the 32x32 "transform matrix"d sprite, and a tad easier on the 64x64 resized one.

If one were to re-sprite every single item which can be put in rechargers (and there is a lot of them), it would be a noble endeavour, but until then we have to rely on transform matrices for those operations. And 64x64 is a gain as far as they're concerned.

The blur is a good thing to mention, now that I think of it. Smaller resolution monitors or people who play in windowed mode have to use byond's stretch-to-fit thing. Which looks a lot worst when it's stretching a 32x32 thing than a 64x64 one, at least, that's what I get from zooming out my screenshots taken in stretch-to-fit in both of these "base resolutions".

As far as holobadge go, the idea is to make the overlay gain a single pixel on each side to make them a bit bigger when worn without increasing their size a lot. Like, currently, it's 3x4 pixels wide, so giving it one pixel of detail to make it bigger would be a 25% increase and it just looks off. 

Basically, the idea was when I was making a worn sprite for a ring and realised it had to be single pixel else it would almost cover the entire hand. There's a lot of tiny things like that.

Another thing someone suggested would be to sprite a different thing when eyes are closed. Currently, the eye is a 1x1 by thing, so it's impossible to do.

You don't need to radically change every sprite, there's a benefit in upscaling without changing the feel of the game. The question of that benefit outweighing the risks and costs is another.

**Comment by Partheo** [comment]
The recharger concept is far from excellent I admit, but I think it shows well enough that a good solution can be found without the necessity of resizing code.

> As far as holobadge go, the idea is to make the overlay gain a single pixel on each side to make them a bit bigger when worn without increasing their size a lot. Like, currently, it's 3x4 pixels wide, so giving it one pixel of detail to make it bigger would be a 25% increase and it just looks off.

I see what you mean. Yes, going to 64x would mean that there's room for more detail. Pretty sure everyone sees eye to eye on that. The amount of detail you could add in the game world after this PR is huge. You could get nicer looking badges, but you can also get nicer looking everything else.

My earlier arguments still stand in saying that, despite this, there is a lot more to lose. Still don't think it's worth it. I'll leave it at that.

**Comment by Kurfursten** [comment]
That's not really an argument at all. Directional sprites are hand drawn because they have to be. You can't do any transformation to a sprite to make it be the backside of something or in profile, short of having a 3D model (in which case you still are making a skin of all sides anyway).  It's bewildering why you would even make a comparison like that, on top of completely misunderstanding what I meant by "ugly from a coding perspective" and inelegant. Before, I thought you knew what you were talking about.

Let me put this another way. Imagine if you had just said, "we have inhand versons of all our guns that are hand drawn, so why don't we stop using overlays entirely and make a version of the human sprite that is holding each gun? Then repeat for every combination of two items held, and then multiply by everything possible to be worn in combination." No... the fact that we have hand drawn directional sprites firmly does not justify that.

**Comment by Partheo** [comment]
> Directional sprites are hand drawn because they have to be. 

I didn't mean every single directional sprite. Of course things like humans need to be hand drawn. I'm not an idiot. But there's plenty of stuff you could just rotate or construct in code- pipes, projectiles, cables, atmos devices, etc. However, having the freedom to draw each direction lets you do cool things like add perspective to a sprite that would otherwise look top-down. This is what I meant.

image

> on top of completely misunderstanding what I meant by "ugly from a coding perspective" and inelegant

Feel free to explain yourself then. Forgive me for believing that beauty from an art perspective is more pertinent than beauty from a code perspective.

> so why don't we stop using overlays entirely 

Why? I never mentioned anything about getting rid of overlays. This whole paragraph doesn't really make sense to me.

If you're referring to the recharger concept, that idea works great with overlays too. Besides there's no combination happening with rechargers, it's always just recharger + 1 gun so there isn't really an issue of having too many states.

>the fact that we have hand drawn directional sprites firmly does not justify that.

Ignoring the fact that you're referring to a "justification" that you made up, what about the other things I mentioned?

 Inhands are hand drawn- you could instead resize and rotate the item sprite and put it into the player's hand with an overlay. No sprites needed! Elegant from a coding perspective, right? Would look like trash in game.

 All the jumpsuit variants, lockers, clothes, etc that come in all sorts of colors are hand drawn- you could instead just have a grayscale sprite and recolor them using code. Elegant. You would lose the ability to add any difference besides color to the different jumpsuits (roboticist comes to mind), and everything would look washed out and oversatured.

Let's find some more examples of poorly maintainable, excessive spriting to cover "everything to have specific sprites for every possible object":

Material sprites, like sheets- instead of having custom designed bars and minerals, you could have a base ingot shape, or a base gem shape, and then apply a texture and color overlay over the sprite. Why draw separate sprites for all the metals when you could just do it in code?

Cocktails- just have a few glasses, and change the color of the liquid inside, easy peasy. No more custom sprites for every single new cocktail. So maintainable, so easy.

Circuit boards- they all look pretty much the same IRL right? So we could have one circuit board sprite, and maybe recolor a little sticker overlay with some code.

---

I could go on. If I am correctly interpreting your stance, you are saying that it is a lot more work to do all these custom sprites. *Yes. It is.* It's a ton more work, and the game looks all the better for it. Making the recharger resize a sprite is the cheap, easy way out. It looks "easy and maintainable," and it also looks like shit. Making a game look good takes effort, and that effort is exactly why we have so many sprites for so many different things. It makes all the difference.

**Comment by Kurfursten** [comment]
This is an entirely bizarre line of argumentation.

"We should have a different recharger sprite for every gun that could be overlayed on it" is exactly the same argument as "we should have a different human sprite for every jumpsuit that could be overlayed on it". It's not at all the same as "two different bar drinks need distinct icons because they are different items". Is this supposed to be convincing?

**Comment by Partheo** [comment]
The only thing that's entirely bizzare is your continued belief that I want to remove overlays. No. The gif I posted earlier with the recharger was accomplished with overlays. You overlay a jumpsuit onto a human sprite. You overlay a gun state onto a recharger sprite. Do you get it?

**Comment by SkowronX** [comment]
cursed

**Comment by Kammerjunk** [comment]
better transforms is interesting
better inhands is a maybe, haven't seen anything about it
different perspectives is haram
higher resolution is iffy

**Comment by Exxion** [comment]
The transforms are mixels though, so they look worse than they already do IMO
Also we haven't seen any screenshots of rotation transforms, only scale ones, and I suspect we'd see some pretty weird artifacting on rotations
To be fair though we only really use rotations in animations that are too fast to notice it

**Comment by Kammerjunk** [comment]
I don't really think they look worse than currently
they look really REALLY bad right now
good point with rotations though

**Comment by ShiftyRail** [comment]
Rotations are ok at a glance. Maybe his shoes look off.

Scaling on some objects might cause eye sore, though, but of a different kind that the one we have current.

rotatate gang

laser canon

**Comment by ShiftyRail** [comment]
There's a persistent BYOND bug and some evident lack of enthusiasm from the community for this to be tried serenely. Maybe at a later time.

image

https://gist.github.com/ShiftyRail/1dbbe757588b89e4a1a01f2cdb6676af 
Instructions for anyone who wish to attempt this. This PR is complete (as far as a single man can test) and merge-to-downstream ready (minus the byond bug) at commit c2a5b4649b6657b9eb5bdaa0c622daae1cae212d.

**Comment by d3athrow** [comment]
😔

**Comment by SonixApache** [comment]
😔

**Comment by SkowronX** [comment]
Womp womp

**Comment by Exxion** [comment]
Shifty, that's not a bug. It's just displaying the sprites at full size, as expected. Full size simply got bigger.

**Comment by ShiftyRail** [comment]
@Exxion There's (supposedly) a way to make it display smaller icons with skin fuckery, but it doesn't work with the `info` panel, only with user-defined grids.

**Comment by patpol4** [comment]
weird
